File: src/main/java/com/miempresa/miproyecto/service/ProductService.java
package com.miempresa.miproyecto.service;

import com.miempresa.miproyecto.model.Product;

import java.util.List;
import java.util.Optional;

public interface ProductService {
    List<Product> getAllProducts();

    Product findProductById(Long id);

    Product createProduct(Product product);

    Product updateProductById(Long id, Product product);

    void deleteProductById(Long id);
}

File: src/app/services/product.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Product } from '../models/product.model';

@Injectable({
  providedIn: 'root',
})
export class ProductService {

  private headers = new HttpHeaders({
    'Authorization': 'Basic ' + btoa('admin:admin123')
  });

  private apiUrl = 'http://localhost:8080/products';


  constructor(private http: HttpClient) {}

  getAllProducts(): Observable<Product[]> {
    return this.http.get<Product[]>(this.apiUrl, { headers: this.headers });
  }

  createProduct(product: Product): Observable<Product> {
    return this.http.post<Product>(this.apiUrl, product, { headers: this.headers });
  }

  updateProductById(id: number, product: Product): Observable<Product> {
    return this.http.put<Product>(`${this.apiUrl}/${id}`, product, { headers: this.headers });
  }

  deleteProductById(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`, { headers: this.headers });
  }
}
